articles/public-key-cryptography-in-js/index.md

```diff
@@ -0,0 +1,235 @@
+
+- **Diffie Hellman protocol**: Diffie Hellman protocol is a mechanism of using the [public channel](https://en.wikipedia.org/wiki/Diffie%E2%80%93Hellman_key_exchange) to exchange cryptographic keys. It was designed by [Whitfield Diffie](https://en.wikipedia.org/wiki/Whitfield_Diffie) and [Martin Hellman](https://en.wikipedia.org/wiki/Martin_Hellman) and was first published in 1976.
+
+### Libraries for public-key cryptography in JavaScript
+There are different libraries for implementing public-key cryptography in JavaScript. The following are the most commonly used.
+
+- [NaCL](http://nacl.cr.yp.to/): It's a high-speed library for carrying out encryption, decryption, and network communication. It uses an elliptic curve cryptography algorithm.
+
+- [TweetNaCL](http://tweetnacl.cr.yp.to/): It was among the initial cryptographic libraries to be released and it was originally written in the C programming language. TweeNaCL.js is the JavaScript version of the library. It uses the Diffie Hellman algorithm.
+
+### Using TweetNaCL.js to implement public-key cryptography
+For this article, we'll use TweetNaCL.js to implement the concept of public-key cryptography.
+
+First, let's install the dependencies to use in the application. You can use `npm` or `yarn`.
+
+```bash
+npm install tweetnacl tweetnacl-util
+```
+
+or 
+
+```bash
+yarn add tweetnacl tweetnacl-util
+```
+
+#### Forming a puzzle 
+Let's come up with a puzzle so that we can implement a solution.
+
+We have two communicating pairs, David and Viktoria. When David is sending a message to Viktoria, he encrypts it using Viktoria's public key. When Viktoria receives it, she decrypts it using her private key. When she decides to reply to the message, she encrypts her message using David's public key, and on David receiving the message, he decrypts it using his private key.
+
+#### Importing libraries
+```JavaScript
+//import the libraries
+const nacl = require('tweetnacl');
+nacl.util = require('tweetnacl-util');
+```
+
+#### Generating the keys
+```JavaScript
+//Generate the keys
+const david = nacl.box.keyPair();
+const viktoria = nacl.box.keyPair();
+```
+
+#### David encrypting the message
+```JavaScript
+function davidEncrypting(){
+    const one_time_code = nacl.randomBytes(24);
+
+    //Get the message from david
+    const plain_text = "Hello there Viktoria";
+
+    //Get the cipher text
+    const cipher_text = nacl.box(
+        nacl.util.decodeUTF8(plain_text),
+        one_time_code,
+        viktoria.publicKey,
+        david.secretKey
+    );
+
+    //message to be sent to Viktoria
+    const message_in_transit = {cipher_text,one_time_code};
+
+    return message_in_transit;
+};
+```
```

Comment:
Can you explain what is a one-time code? and Explain what the decodeUTF8 function does?

Reply:
resolved